### PR TITLE
Fixed #33252 -- Made cache middleware thread safe.

### PR DESCRIPTION
**Pull request by iurisilvio**
The `caches[key]` response is not thread safe. It must be called inside
the correct thread to return a new object.

https://docs.djangoproject.com/en/3.2/topics/cache/#accessing-the-cache

### CODE REVIEW

File: django/middleware/cache.py [hunk 1]
```diff
@@ -67,7 +67,17 @@ def __init__(self, get_response):
...
-        self.cache = caches[self.cache_alias]
+
+    @property
+    def cache(self):
+        try:
+            return self._cache
+        except AttributeError:
+            return caches[self.cache_alias]
+
+    @cache.setter
```
Comment (iurisilvio) [inline django/middleware/cache.py:78]: I don't like this setter, but some tests update cache attribute.

**Comment by iurisilvio** [comment]
I added a second commit removing the setter that I added in the first one and changing other tests to make it pass. I don't know if I'm allowed to do that.

The setter I added first was to make other tests pass, but removing this set don't break anything.

File: tests/cache/tests.py [hunk 2]
```diff
@@ -1011,7 +1010,6 @@ def get_response(req):
             return response
 
         update_middleware = UpdateCacheMiddleware(get_response)
-        update_middleware.cache = cache
```
Comment (felixxm) [inline tests/cache/tests.py:1014]: `FetchFromCacheMiddleware` and `UpdateCacheMiddleware` don't use `default` cache by default. We should override `CACHE_MIDDLEWARE_ALIAS` to make sure they will use the same cache as before. We can do this in a separate commit. For example:
```diff
diff --git a/tests/cache/tests.py b/tests/cache/tests.py
index 29e527d4e6..13ff0a45aa 100644
--- a/tests/cache/tests.py
+++ b/tests/cache/tests.py
@@ -994,9 +994,9 @@ class BaseCacheTests:
         self.assertEqual(caches['custom_key'].get('answer2'), 42)
         self.assertEqual(caches['custom_key2'].get('answer2'), 42)
 
+    @override_settings(CACHE_MIDDLEWARE_ALIAS=DEFAULT_CACHE_ALIAS)
     def test_cache_write_unpicklable_object(self):
         fetch_middleware = FetchFromCacheMiddleware(empty_response)
-        fetch_middleware.cache = cache
 
         request = self.factory.get('/cache/test')
         request._cache_update_cache = True
@@ -1011,7 +1011,6 @@ class BaseCacheTests:
             return response
 
         update_middleware = UpdateCacheMiddleware(get_response)
-        update_middleware.cache = cache
         response = update_middleware(request)
 
         get_cache_data = fetch_middleware.process_request(request)

```

File: tests/cache/tests.py [hunk 3]
```diff
@@ -2489,6 +2487,24 @@ def test_304_response_has_http_caching_headers_but_not_cached(self):
         self.assertIn('Cache-Control', response)
         self.assertIn('Expires', response)
 
+    def test_per_thread(self):
+        """
+        Cache object must be different for each thread.
+        """
```
Comment (felixxm) [inline tests/cache/tests.py:2493]:
```suggestion
        """The cache instance is different for each thread."""
```

File: tests/cache/tests.py (hunk 2 above)
Comment (iurisilvio) [reply tests/cache/tests.py:1014]: I used your suggestion. I could change `middlewaare.cache_alias` instead of `middleware.cache` to avoid the property issue, but I preferred your solution.

**Comment by felixxm** [comment]
@iurisilvio Thanks :+1:

File: tests/cache/tests.py [hunk 4]
```diff
@@ -2489,6 +2488,21 @@ def test_304_response_has_http_caching_headers_but_not_cached(self):
         self.assertIn('Cache-Control', response)
         self.assertIn('Expires', response)
 
+    def test_per_thread(self):
+        """The cache instance is different for each thread."""
```
Comment (pauloxnet) [inline tests/cache/tests.py:2492]: I know this is just a test, but to increase future readability I would suggest using talking names and not single characters (e.g. `c`, `t`, `x`) to store variables, and I would also remove unnecessary blank lines

Comment (felixxm) [reply tests/cache/tests.py:2492]: I updated variable names.

> and I would also remove unnecessary blank lines

I don't see any unnecessary blank lines :thinking:

Comment (pauloxnet) [reply tests/cache/tests.py:2492]: I'm sorry, the short variables had misled me, in fact there aren't any. Thanks

**Review by pauloxnet — APPROVED** [review]

File: tests/cache/tests.py (hunk 4 above)
Comment (iurisilvio) [reply tests/cache/tests.py:2492]: I copied that from a similar test, but I agree it is not a good reason to continue with a bad practice. Thanks for the corrections!
